### PR TITLE
PIN-2204 - [BE] La lista delle finalità visibili in un client deve escludere quelle in stato ARCHIVED

### PR DESCRIPTION
**Pull request by nttdata-rtorsoli**

### CODE REVIEW

File: src/main/scala/it/pagopa/interop/authorizationprocess/api/impl/ClientApiServiceImpl.scala [hunk 1]
```diff
@@ -630,4 +630,27 @@ final case class ClientApiServiceImpl(
...
+  )(implicit contexts: Seq[(String, String)], toEntityMarshallerProblem: ToEntityMarshaller[Problem]): Route =
+    authorize(ADMIN_ROLE) {
+      val operationLabel: String = s"Removing archived Purpose $purposeId from all clients"
+      logger.info(operationLabel)
+
+      val result: Future[Unit] = for {
+        purposeUuid <- purposeId.toFutureUUID
+        purpose     <- purposeManagementService.getPurpose(purposeUuid)
+        _           <- purpose.versions
+          .find(_.state == PurposeManagementDependency.PurposeVersionState.ARCHIVED)
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/authorizationprocess/api/impl/ClientApiServiceImpl.scala:645]: a purpose can be considered ARCHIVED if all its versions are archived

File: src/main/scala/it/pagopa/interop/authorizationprocess/common/readmodel/ReadModelQueries.scala [hunk 2]
```diff
@@ -110,6 +110,19 @@ object ReadModelQueries {
...
+    purposeId: UUID
+  )(readModel: ReadModelService)(implicit ec: ExecutionContext): Future[Seq[PersistentClient]] = {
+
+    val query: Bson = Filters
+      .and(Filters.eq("data.purposes.purpose.purposeId", purposeId.toString))
+
+    for {
+      clients <- readModel
+        .find[PersistentClient]("clients", query, offset = 0, limit = Int.MaxValue)
+    } yield clients
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/authorizationprocess/common/readmodel/ReadModelQueries.scala]:
```suggestion
    readModel.find[PersistentClient]("clients", query, offset = 0, limit = Int.MaxValue)
```

File: src/test/scala/it/pagopa/interop/authorizationprocess/util/SpecUtils.scala [hunk 3]
```diff
@@ -237,6 +271,18 @@ trait SpecUtils extends SprayJsonSupport { self: MockFactory =>
...
+  val archivedPurpose: PurposeManagementDependency.Purpose =
+    purpose.copy(
+      id = UUID.randomUUID(),
+      versions = Seq(purposeVersion.copy(state = PurposeManagementDependency.PurposeVersionState.ARCHIVED))
+    )
+
+  val notArchivedPurpose: PurposeManagementDependency.Purpose =
+    purpose.copy(
+      id = UUID.randomUUID(),
+      versions = Seq(purposeVersion.copy(state = PurposeManagementDependency.PurposeVersionState.ACTIVE))
```
Comment (galales) [inline src/test/scala/it/pagopa/interop/authorizationprocess/util/SpecUtils.scala]: I suggest to add this case to improve testing
```suggestion
      versions = Seq(purposeVersion.copy(state = PurposeManagementDependency.PurposeVersionState.ACTIVE), purposeVersion.copy(state = PurposeManagementDependency.PurposeVersionState.ARCHIVED))
```

File: src/main/scala/it/pagopa/interop/authorizationprocess/api/impl/ClientApiServiceImpl.scala [hunk 4]
```diff
@@ -630,4 +630,27 @@ final case class ClientApiServiceImpl(
...
+          .toFuture(PurposeNoVersionFound(purpose.id))
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/authorizationprocess/api/impl/ClientApiServiceImpl.scala]: so this should be more something like `PurposeNotInExpectedState`

**Review by galales — CHANGES_REQUESTED** [review]

File: src/test/scala/it/pagopa/interop/authorizationprocess/util/SpecUtils.scala (hunk 3 above)
Comment (nttdata-rtorsoli) [reply src/test/scala/it/pagopa/interop/authorizationprocess/util/SpecUtils.scala]: But it must be not in Archived STATE, in this case. Maybe I need to isolate the latest version?

File: src/main/scala/it/pagopa/interop/authorizationprocess/api/impl/ClientApiServiceImpl.scala (hunk 1 above)
Comment (nttdata-rtorsoli) [reply src/main/scala/it/pagopa/interop/authorizationprocess/api/impl/ClientApiServiceImpl.scala:645]: Please tell me What do you want to do in this case.

Comment (nttdata-rtorsoli) [reply src/main/scala/it/pagopa/interop/authorizationprocess/api/impl/ClientApiServiceImpl.scala:645]: //All are archived
`purpose.versions
        .filterNot(_.state == PurposeManagementDependency.PurposeVersionState.ARCHIVED)
        .maxByOption(_.createdAt)
        .toFuture(PurposeNotInExpectedState(purpose.id))`
          
//The latest is archived
 `purpose.versions
         .maxByOption(_.createdAt)
          .find(_.state == PurposeManagementDependency.PurposeVersionState.ARCHIVED)
          .toFuture(PurposeNotInExpectedState(purpose.id))`

Which is the correct answer?

**Review by galales — APPROVED** [review]

**Review by beetlecrunch — APPROVED** [review]

**Review by TonioGela — APPROVED** [review]

**Pull request merged**